db/limitersMgr.go

```diff
@@ -0,0 +1,696 @@
+}
+
+func getLimiterKey(limiterName string) []byte {
+	var key []byte
+	key = append(key, []byte(LIMITER_STATUS_PREFIX)...)
+	key = append(key, []byte(limiterName)...)
+	key = append(key, ':')
+	return key
+}
+
+func getNamespaceAndCmd(limiterName string) []string {
+	strs := strings.Split(limiterName, NAMESPACE_COMMAND_TOKEN)
+	if len(strs) < 2 {
+		return nil
+	}
+	return strs
+
+}
+
+func NewLimitersMgr(store *RedisStore, rateLimit *conf.RateLimit) (*LimitersMgr, error) {
+	var addrs []net.Addr
+	var err error
+	if rateLimit.InterfaceName != "" {
+		iface, err := net.InterfaceByName(rateLimit.InterfaceName)
+		if err != nil {
+			return nil, err
+		}
+
+		addrs, err = iface.Addrs()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		addrs, err = net.InterfaceAddrs()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	localIp := ""
+	for _, a := range addrs {
+		if ipnet, ok := a.(*net.IPNet); ok && !ipnet.IP.IsLoopback() && ipnet.IP.To4() != nil {
+			localIp = ipnet.IP.String()
+			break
+		}
+	}
+	if localIp == "" {
+		return nil, errors.New(rateLimit.InterfaceName + " adds is empty")
+	}
+
+	if rateLimit.LimiterNamespace == "" {
+		return nil, errors.New("limiter-namespace is configured with empty")
+	}
+	if rateLimit.WeightChangeFactor <= 1 {
+		return nil, errors.New("weight-change-factor should > 1")
+	}
+	if !(rateLimit.UsageToDivide > 0 && rateLimit.UsageToDivide < rateLimit.UsageToMultiply && rateLimit.UsageToMultiply < 1) {
+		return nil, errors.New("should config 0 < usage-to-divide < usage-to-multiply < 1")
+	}
+	if rateLimit.InitialPercent > 1 || rateLimit.InitialPercent <= 0 {
+		return nil, errors.New("initial-percent should in (0, 1]")
+	}
+
+	l := &LimitersMgr{
+		limitDatadb: store.DB(rateLimit.LimiterNamespace, LIMITDATA_DBID),
+		conf:        rateLimit,
+		localIp:     localIp,
+	}
+
+	go l.startSyncNewLimit()
+	go l.startReportAndBalance()
+	return l, nil
+}
+
+func (l *LimitersMgr) init(limiterName string) *CommandLimiter {
+	//lock is just prevent many new connection of same namespace to getlimit from tikv in same time
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	v, ok := l.limiters.Load(limiterName)
+	if ok {
+		return v.(*CommandLimiter)
+	}
+
+	allmatchLimiterName := getAllmatchLimiterName(limiterName)
+	l.qpsAllmatchLimit.LoadOrStore(allmatchLimiterName, (*LimitData)(nil))
+	l.rateAllmatchLimit.LoadOrStore(allmatchLimiterName, (*LimitData)(nil))
+
+	qpsLimit, qpsBurst := l.getLimit(limiterName, true)
+	rateLimit, rateBurst := l.getLimit(limiterName, false)
+	if (qpsLimit > 0 && qpsBurst > 0) ||
+		(rateLimit > 0 && rateBurst > 0) {
+		newCl := NewCommandLimiter(l.localIp, limiterName, qpsLimit, qpsBurst, rateLimit, rateBurst, l.conf.InitialPercent)
+		v, _ := l.limiters.LoadOrStore(limiterName, newCl)
+		return v.(*CommandLimiter)
+	} else {
+		l.limiters.LoadOrStore(limiterName, (*CommandLimiter)(nil))
+		return nil
+	}
+}
+
+func (l *LimitersMgr) getLimit(limiterName string, isQps bool) (int64, int) {
```

Comment:
Avoid using a boolean as an argument. Refactor it to `getQPSLimit` and `getRateLimit`.